src/cita/indexer.ts

```diff
@@ -0,0 +1,281 @@
+const limiter = new Bottleneck({
+	minTime: 20, // 50 requests per second
+});
+
+//export type OpenAlexID = `W${number}`;
+export type UID =
+	| { DOI: string }
+	| { ISBN: string }
+	| { arXiv: string }
+	| { openAlex: string }
+	| { semantic: string }
+	| { OMID: string }
+	| { adsBibcode: string }
+	| { PMID: string }
+	| { PMCID: string };
+
+export interface IndexedWork<Ref> {
+	referenceCount: number;
+	referencedWorks: Ref[];
+}
+
+export interface LookupIdentifier {
+	type: PIDType;
+	id: string;
+}
+
+export abstract class IndexerBase<Ref> {
+	/**
+	 * Name of the indexer to be displayed.
+	 */
+	abstract indexerName: string;
+
+	/**
+	 * Supported PIDs for the indexer.
+	 */
+	abstract supportedPIDs: PIDType[];
+
+	/**
+	 * Extract supported PID from the source item.
+	 * @param item Source item to extract the PID from.
+	 * @returns Supported PID or null if none
+	 */
+	extractSupportedPID(item: SourceItemWrapper): LookupIdentifier | null {
+		for (const pid of this.supportedPIDs) {
+			const value = item.getPID(pid, true); // Already clean them up
+			if (value) return { type: pid, id: value };
+		}
+
+		return null;
+	}
+
+	/**
+	 * Abstract method to get references from the specific indexer.
+	 * @param {LookupIdentifier[]} identifiers - List of DOIs or other identifiers.
+	 * @returns {Promise<IndexedWork[]>} Corresponding works.
+	 */
+	abstract getReferences(
+		identifiers: LookupIdentifier[],
+	): Promise<IndexedWork<Ref>[]>;
+
+	/**
+	 * Abstract method to parse a list of references into Zotero items.
+	 * @param {Ref[]} references - References for a specific work.
+	 * @returns {Promise<Zotero.Item[]>} Zotero items parsed from the references.
+	 */
+	abstract parseReferences(references: Ref[]): Promise<Zotero.Item[]>;
+
+	/**
+	 * Filter source items with supported UIDs.
+	 * @param sourceItems Selected items to filter depending on the indexer.
+	 */
+	filterItemsWithSupportedIdentifiers(
+		sourceItems: SourceItemWrapper[],
+	): [sourceItems: SourceItemWrapper[], identifiers: LookupIdentifier[]] {
+		const identifiers: LookupIdentifier[] = [];
+		const filteredItems: SourceItemWrapper[] = [];
+
+		for (const item of sourceItems) {
+			const uid = this.extractSupportedPID(item);
+			if (uid) {
+				identifiers.push(uid);
+				filteredItems.push(item);
+			}
+		}
+
+		return [filteredItems, identifiers];
+	}
+
+	/**
+	 * Get source item citations from the online database.
+	 * @param {SourceItemWrapper[]} sourceItems - One or more source items to get citations for.
+	 */
+	async addCitationsToItems(
+		sourceItems: SourceItemWrapper[],
+		autoLinkCitations = true,
+	) {
+		// Filter items with valid identifiers (DOI or other)
+		const [fetchableSourceItems, identifiers] =
+			this.filterItemsWithSupportedIdentifiers(sourceItems);
+		if (fetchableSourceItems.length === 0) {
+			Services.prompt.alert(
+				window as mozIDOMWindowProxy,
+				Wikicite.formatString(
+					"wikicite.indexer.get-citations.no-doi-title",
```

Comment:
Shouldn't these messages say: "No items with a supported identifer provided" found instead of "No items with a DOI provided"?

This message could also then also have the list of supported identifiers?